R/join_keys.R

```diff
@@ -0,0 +1,651 @@
+    dataset_1 <- names(join_keys_obj)[[dataset_1]]
+  }
+
+  checkmate::assert_string(dataset_1)
+
+  # Accepting 1 subscript with valid `value` formal
+  checkmate::assert_list(value, names = "named", types = "character", null.ok = TRUE)
+
+  # Normalize values
+  norm_value <- lapply(names(value), function(.x) {
+    get_keys(join_key(dataset_1, .x, value[[.x]]))
+  })
+
+  names(norm_value) <- names(value)
+  value <- norm_value
+
+  #
+  # Remove classes to use list-based get/assign operations
+  join_keys_obj <- unclass(join_keys_obj)
+
+  # In case a pair is removed, also remove the symmetric pair
+  removed_names <- setdiff(names(join_keys_obj[[dataset_1]]), names(value))
+  if (length(removed_names) > 0) {
+    for (.x in removed_names) join_keys_obj[[.x]][[dataset_1]] <- NULL
+  }
+
+  join_keys_obj[[dataset_1]] <- value
+
+  # Iterate on all new values to create symmetrical pair
+  for (ds2 in names(value)) {
+    if (ds2 == dataset_1) next
+
+    keep_value <- join_keys_obj[[ds2]] %||% list()
+    new_value <- value[[ds2]]
+
+    if (checkmate::test_character(new_value, min.len = 1, names = "unnamed")) {
+      new_value <- setNames(new_value, new_value)
+    } else if (
+      checkmate::test_character(new_value, min.len = 1)
+    ) {
+      # Invert key
+      new_value <- setNames(names(new_value), new_value)
+    }
+
+    keep_value[[dataset_1]] <- new_value
+
+    # Assign symmetrical
+    join_keys_obj[[ds2]] <- keep_value
+  }
+
+  # Remove NULL or empty keys
+  empty_ix <- vapply(
+    join_keys_obj,
+    function(.x) is.null(.x) || length(.x) == 0,
+    logical(1)
+  )
+  preserve_attr <- attributes(join_keys_obj)[!names(attributes(join_keys_obj)) %in% "names"]
+  join_keys_obj <- join_keys_obj[!empty_ix]
+  attributes(join_keys_obj) <- modifyList(attributes(join_keys_obj), preserve_attr)
+
+  #
+  # restore class
+  class(join_keys_obj) <- c("join_keys", "list")
+  join_keys_obj
+}
+
+#' @rdname merge_join_keys
+#' @keywords internal
+merge_join_keys <- function(join_keys_obj, new_join_keys) {
+  UseMethod("merge_join_keys", join_keys_obj)
+}
+
+#' @rdname merge_join_keys
+#' @keywords internal
+merge_join_keys.default <- function(join_keys_obj, new_join_keys) {
+  merge_join_keys(join_keys(join_keys_obj), new_join_keys)
+}
+
+#' Merging a list (or one) of `join_keys` objects into the current `join_keys` object
+#'
+#' @rdname merge_join_keys
+#'
+#' @param join_keys_obj (`join_keys`) object to merge the new_join_keys.
+#' @param new_join_keys  `list` of `join_keys` objects or single `join_keys` object
+#'
+#' @return a new `join_keys` object with the resulting merge.
+#'
+#' @keywords internal
+merge_join_keys.join_keys <- function(join_keys_obj, new_join_keys) {
+  if (
+    checkmate::test_class(new_join_keys, "join_key_set") ||
+      checkmate::test_class(new_join_keys, "join_keys")
+  ) {
+    new_join_keys <- list(new_join_keys)
+  }
+
+  lapply(new_join_keys, assert_join_keys_alike)
```

Comment:
This seems redundant to me. Why `test` for class membership, wrap in `list`, and then `lapply` an assertion on class membership?

Reply:
The assertion was related to structure, not really class membership.

Moot point as it was sinced removed.